make faster BigFloats

vtjnash opened the pull request:
We can coalesce the two required allocations for the MFPR BigFloat API design into one allocation, hopefully giving a easy performance boost. It would have been slightly easier and more efficient if MPFR BigFloat was already a VLA instead of containing a pointer here, but that does not prevent the optimization.

@nanosoldier `runbenchmarks("BigFloat", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/55906) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/33e724b_vs_3aad027/report.md).

giordano commented:
This seems to broken on 32-bit platforms

vtjnash commented:
Oof, 3GB logs. Looks like a slight miscomputation of the length(Limbs). I fixed this on 64-bit, but might not have gotten the math right to be the same on 32-bit
```
/cache/build/tester-amdci4-12/julialang/julia-master/julia-36ee135d97/share/julia/test/rounding.jl:424
Expression: mpfr_to_ieee(F, bf, mpfr_rm) === F(bf, mpfr_rm) === F(bf, convert(RoundingMode, mpfr_rm))
Evaluated: 1.086461844974e-311 === 1.0864618452707e-311 === 1.0864618452707e-311
```

vtjnash committed:
make faster BigFloats

We can coalesce the two required allocations for the MFPR BigFloat API
design into one allocation, hopefully giving a easy performance boost.
It would have been slightly easier and more efficient if MPFR BigFloat
was already a VLA instead of containing a pointer here, but that does
not prevent the optimization.

vtjnash commented:
Okay, 32-bit turned out to be an existing bug (OOB read) that got more easily detected now since (a) we aren't doing unsafe_read anymore (b) the read previously happened to land in the length field of a length 4 string, and now lands in a pointer value instead

Keno commented:
Seeing this on CI:
```

julia: /cache/build/builder-amdci5-4/julialang/julia-master/src/gc-stock.c:794: jl_gc_small_alloc_inner: Assertion `pg->osize == p->osize' failed.
--
  |  
  | [10851] signal 6 (-6): Aborted
  | in expression starting at /cache/build/tester-amdci4-12/julialang/julia-master/julia-6b9719f767/share/julia/test/gc/chunks.jl:13
  | gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
  | abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
  | unknown function (ip: 0x7fd248b4540e)
  | __assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
  | jl_gc_small_alloc_inner at /cache/build/builder-amdci5-4/julialang/julia-master/src/gc-stock.c:794 [inlined]
  | jl_gc_small_alloc_inner at /cache/build/builder-amdci5-4/julialang/julia-master/src/gc-stock.c:749
  | jl_gc_small_alloc_noinline at /cache/build/builder-amdci5-4/julialang/julia-master/src/gc-stock.c:818 [inlined]
  | jl_gc_alloc_ at /cache/build/builder-amdci5-4/julialang/julia-master/src/julia_internal.h:528 [inlined]
  | jl_gc_alloc at /cache/build/builder-amdci5-4/julialang/julia-master/src/gc-stock.c:3404
  | _new_genericmemory_ at /cache/build/builder-amdci5-4/julialang/julia-master/src/genericmemory.c:56
  | GenericMemory at ./boot.jl:538 [inlined]
  | _ at ./mpfr.jl:151 [inlined]
  | BigFloat at ./mpfr.jl:147 [inlined]
  | rand at /cache/build/builder-amdci5-4/julialang/julia-master/usr/share/julia/stdlib/v1.12/Random/src/generation.jl:103
```
Haven't debugged it yet, so no reason to think it's related to this PR other than that it's in this code and showed up recently.

https://buildkite.com/julialang/julia-master/builds/40519#019252fb-3c0e-47c0-9d56-96e43fd6d394

blegat commented:
This started failing on nightly, probably because of this PR:
```julia
a = one(BigFloat)
b = zero(BigFloat)
ccall((:mpfr_swap, :libmpfr), Cvoid, (Ref{BigFloat}, Ref{BigFloat}), b, a)
show(b)
```
It gives
```
get_str.c:157: MPFR assertion failed: size_s1 >= m

[42807] signal 6 (-6): Aborted
in expression starting at /home/blegat/.julia/dev/MutableArithmetics/b.jl:6
unknown function (ip: 0x725fea6533f4)
gsignal at /usr/lib/libc.so.6 (unknown line)
abort at /usr/lib/libc.so.6 (unknown line)
mpfr_assert_fail at /workspace/srcdir/mpfr-4.2.1/src/mpfr-gmp.c:300
mpfr_get_str_aux at /workspace/srcdir/mpfr-4.2.1/src/get_str.c:157
mpfr_get_str at /workspace/srcdir/mpfr-4.2.1/src/get_str.c:2877
regular_eg at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:1308
partition_number at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:1925 [inlined]
sprnt_fp at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:1994 [inlined]
mpfr_vasnprintf_aux at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:2413
mpfr_asprintf at /workspace/srcdir/mpfr-4.2.1/src/printf.c:176
string_mpfr at ./mpfr.jl:1187
_string at ./mpfr.jl:1233
_string at ./mpfr.jl:1235 [inlined]
show at ./mpfr.jl:1245
unknown function (ip: 0x725fe8d03866)
show at ./show.jl:476
unknown function (ip: 0x725fe8d00d62)
jl_apply at /cache/build/builder-amdci5-1/julialang/julia-master/src/julia.h:2228 [inlined]
do_call at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:125
eval_value at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:222
eval_stmt_value at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:173 [inlined]
eval_body at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:685
jl_interpret_toplevel_thunk at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:892
jl_toplevel_eval_flex at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:1047
jl_toplevel_eval_flex at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:987
ijl_toplevel_eval at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:1058
ijl_toplevel_eval_in at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:1100
eval at ./boot.jl:439 [inlined]
include_string at ./loading.jl:2741
_include at ./loading.jl:2801
include at ./Base.jl:581 [inlined]
exec_options at ./client.jl:329
_start at ./client.jl:558
jfptr__start_66271.1 at /home/blegat/.julia/juliaup/julia-nightly/lib/julia/sys.so (unknown line)
jl_apply at /cache/build/builder-amdci5-1/julialang/julia-master/src/julia.h:2228 [inlined]
true_main at /cache/build/builder-amdci5-1/julialang/julia-master/src/jlapi.c:900
jl_repl_entrypoint at /cache/build/builder-amdci5-1/julialang/julia-master/src/jlapi.c:1059
main at /cache/build/builder-amdci5-1/julialang/julia-master/cli/loader_exe.c:58
unknown function (ip: 0x725fea5e2e07)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8)
Allocations: 1 (Pool: 1; Big: 0); GC: 0
```
See https://github.com/jump-dev/MutableArithmetics.jl/pull/307

vtjnash commented:
That has always been UB, since it moved the data but not the root, it might randomly segfault later or suddenly switch back to the old value

odow commented:
Agreed. I've fixed this in our downstream code. Our original implementation was needlessly complicated.
